### PR TITLE
IgnoreBaseline widget

### LINKED ISSUE #7037: "Ignore for baseline purposes" widget
We should have a widget that causes an entire subtree to be ignored for the purposes of baseline computations.

For example, we would use this to wrap the label of an `InputContainer`.

### PR DESCRIPTION
**Pull request by Hixie**
Fixes https://github.com/flutter/flutter/issues/7037

### CODE REVIEW

**Comment by Hixie** [comment]
cc @justinmc - you might want to use this around the labels in a text field to prevent buttons in the same row from aligning with the text field label rather than the text field contents

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 1]
```diff
@@ -842,6 +842,19 @@ class RenderIntrinsicHeight extends RenderProxyBox {
   }
 }
 
+/// Excludes the child from baseline computations in the parent.
+class RenderIgnoreBaseline extends RenderProxyBox {
+  /// Create a render object that causes the parent to ignore the child for baseline computations.
+  RenderIgnoreBaseline({
+    RenderBox? child,
```
Comment (justinmc) [inline packages/flutter/lib/src/rendering/proxy_box.dart:849]: Can this be `super.child`?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -3561,6 +3559,25 @@ class Baseline extends SingleChildRenderObjectWidget {
...
+    super.key,
+    super.child,
+  });
+
+  @override
+  RenderIgnoreBaseline createRenderObject(BuildContext context) {
+    return RenderIgnoreBaseline();
+  }
+}
+
```
Comment (justinmc) [inline packages/flutter/lib/src/widgets/basic.dart:3580]: Nit: Too many newlines here.

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1: 

Is this what you mean about aligning with the content instead of the label?  Looks like it works now but I'm probably misunderstanding, or maybe we're already disregarding labels etc. in baseline calculations.

<details>

<summary>Code</summary>

```dart
import 'package:flutter/material.dart';

// See https://github.com/flutter/flutter/pull/131220

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController _controller = TextEditingController(
    //text: 'Hello text field.',
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Row(
              crossAxisAlignment: CrossAxisAlignment.baseline,
              textBaseline: TextBaseline.alphabetic,
              textDirection: TextDirection.ltr,
              children: [
                ConstrainedBox(
                  constraints: BoxConstraints.loose(const Size(150.0, double.infinity)),
                  child: TextField(
                    controller: _controller,
                    decoration: const InputDecoration(
                      labelText: 'label',
                      labelStyle: TextStyle(fontSize: 64.0),
                    )
                  ),
                ),
                const Text(
                  'a',
                  textDirection: TextDirection.ltr,
                  style: TextStyle(fontSize: 128.0, fontFamily: 'FlutterTest'), // places baseline at y=96
                ),
                const Text(
                  'b',
                  textDirection: TextDirection.ltr,
                  style: TextStyle(fontSize: 32.0, fontFamily: 'FlutterTest'), // 24 above baseline, 8 below baseline
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

Screenshot from 2023-07-25 10-42-16
Screenshot from 2023-07-25 10-42-00

**Comment by Hixie** [comment]
> Is this what you mean about aligning with the content instead of the label? Looks like it works now but I'm probably misunderstanding, or maybe we're already disregarding labels etc. in baseline calculations.

It was what I meant. I wonder how we make it work today! It used to be broken (see comments in https://github.com/flutter/flutter/issues/7037).

File: packages/flutter/lib/src/widgets/basic.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/widgets/basic.dart:3580]: it's a section separator (the next section starts with an all-caps heading), the rest of the file (with one exception i can fix) has the same double blank line before headings.

I don't feel strongly one way or the other if you would prefer I removed the double blank lines everywhere in this file before the headings, what do you think?

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 1 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/proxy_box.dart:849]: unfortunately not, it's a positional argument in the superclass

Comment (justinmc) [reply packages/flutter/lib/src/rendering/proxy_box.dart:849]: Ah I missed that it was positional.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 2 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/basic.dart:3597]: I say don't worry about it here, maybe if it comes up again we can make this file all 1 line.

**Comment by justinmc** [comment]
I'll post in the issue about the button-textfield alignment thing.

**Pull request merged**